notes/dns-client-proxy.md

```diff
@@ -0,0 +1,667 @@
+
+To future proof applications, there is a single flag DD, that by default
+disallows transports that are not explicitly listed. With this flag clear,
+the application allows all transports that are not explicitly disallowed
+(including future transports). With the flag set, the
+application has to explicitly list which transports can be used.
+For example, by setting only DD and AT, the application forces the
+use of DoT.
+
+When DD = 0:
+
+* all transports are in the pool of potentially usable transports
+* D53, DT, DH2, DH3 and DQ remove those transports from the pool
+
+When DD = 1:
+
+* no transports are in the pool of potentially usable transports
+* A53, AT, AH2, AH3 and AQ add those transports to the pool
+
+
+Finally, an application can specify its own resolvers or rely on the
+resolvers that are known to the proxy. If ADN Length and Addr Length
+are both zero, then the application requests the resolvers known to
+the proxy. [Note: it is unclear at the moment what to do with any
+Service Parameters]
+
+If the application specifies only an authentication-domain-name then the
+proxy is expected to resolve the name to addresses. If only addresses
+are specified then the proxy assumes that no name is known (though
+a PKIX certificate may include an address literal in the subjectAltName).
+If both a name and address are specified then the proxy will use the
+specified address and use the name for authentication.
+
+To simplify the encoding of the option, an option with addresses will
+have either IPv4 or IPv6 addresses. If the application wants to specify
+both IPv4 and IPv6 addresses for a certain authentication-domain-name
+then it has to include two options. 
+
+An application may want to specify a DNS resolver that is reachable 
+through an IPv6 link-local address.
+IPv6 link-local addresses are special in that they require a zone to be
+specified, either explicitly or implicitly. Typically for a link-local
+address that appears as a source or destination address,
+the zone is implicitly the zone of the link the packet travels on.
+For packets that travel between hosts, there is no goed way to explictly
+specify the zone of a link-local address because two different hosts
+do not agree on zone names. However, if the proxy is on the same host
+as the application, then the zone identifier for the link-local address
+can be specified in the Interface field. For this purpose an interface
+name can also be an interface index expressed as a decimal string.
+
+
+When present, Service Parameters specify how to connect. Otherwise it is
+up to the proxy to try various possibilities. For Service Parameters,
+the values of the ipv4hint and ipv6hint fields are ignored. Addresses
+can only be specified using the addresses field in the PROXY CONTROL Option.
+
+Associated with this option is a new error, BADPROXYPOLICY. When
+a proxy cannot meet the requirements in a PROXY CONTROL Option or the
+option is malformed, it returns this error.
```

Comment:
Before this also contained text that stated that the error was returned " along with a PROXY CONTROL option that lists what it can do.". Do we no longer want that? I think it is said before that the PROXY CONTROL option can also be used to report on capabilities, and also to probe for capabilities.

Reply:
Yes, I removed it. I think it is too confusing to use such a value in an application. If it has a lot of diagnostic value, then we could make it optional for the proxy to return this.

Just trying to keep it simple and remove mandated complexity.